Fix #5012: Remove KitKat support

BenHenning opened the pull request:
## Explanation
Fixes #5012

Removes references and code corresponding to KitKat (SDK version 19) since the app has been set to a minimum version of Lollipop (SDK 21) since #3910 (roughly--technically KitKat was still supported but we no longer shipped a KitKat version). Because of this, the changes in this PR are largely a no-op from a build perspective.

More broadly, this change is motivated by a desire to decrease CI time (which was already reduce considerably in the recent merge of #5629) by removing extraneous app builds being done in CI:
- ``//:oppia`` since we should really only use ``//:oppia_dev`` moving forward.
- Dev and alpha KitKat builds (the latter of which takes a long time since only the alpha job was building 2 proguarded builds of the app and thus taking much more time than the beta and GA build jobs).

Separately, ``build_tests.yml`` and ``unit_tests.yml`` were updated to no longer support caching since we disabled this behavior a while back (#2884) and are unlikely to reintroduce it due to high storage costs.

Some other noteworthy changes:
- The optional providing of ``Retrofit`` and Retrofit services has been reverted since it's no longer necessary (and corresponding tests have been removed since there's no longer any condition to verify).
- Code that was gated to run below Lollipop has been removed since it can't execute except on KitKat devices.
- Support for a main dex target list has been removed as we're unlikely to need it with native multidex support (which wasn't an option for KitKat builds), though we may choose to reintroduce it to speed up cold app starts since it _can potentially_ help improve time-to-splash app startup performance.
- Updated manifest min SDK values to avoid potential developer confusion on the min SDK version supported (though, strictly speaking, this doesn't technically matter with the way the app builds).
- This updates the default min version supported for OS deprecation. No additional work is needed to actually activate the deprecation notice on the KitKat version of the app since we no longer deploy it, and OS deprecation wasn't added to the last version that was deployed.
- Three file content pattern failure messages were updated to no longer reference KitKat (though the value of these checks mostly still seems realized, so I think it's fine to keep them in).
- Some additional licenses were added due to tooling around the dev AAB (even though these aren't dependencies that ship with the app). I think it's fine including extra licenses in this case.
- The protobuf license was corrected to BSD 3-Clause as declared for the dependency and in the GitHub repository for the dependencies.
- This bumps version codes due to removing two flavors (for KitKat dev and KitKat alpha).
- ``NetworkModuleTest`` has had its tests redesigned and better filled in in order to pass the code coverage minimum requirement (since old tests were removed). Note that it's not quite perfect in what it's testing, but it is at least verifying the complete configuration of the ``Retrofit`` instance, and the singleton properties of the two services currently supported (verifying that the services are set up would require a lot more testing that seems outside the direct scope of this test, so it seems okay to ignore here).
- As part of the previous item's work, ``NetworkLoggingInterceptor`` had an issue fixed where it could cause an OkHttp exception to be thrown when trying to process logs (due to reading the response body twice). Reverting this change will now cause breakage failures in ``NetworkModuleTest``.
- ``NetworkModule``'s initialization order was changed for slightly better network request logging (see the new comment in that file for more context).

Note that there are some workflow job removals and renames in this PR. Any required CI checks will be correspondingly updated once this PR is approved and ready to merge (to avoid blocking any other in-flight PRs).

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
This shouldn't have an side effects on non-KitKat UIs, and KitKat support itself is being removed in this PR (so there's nothing to demonstrate).

BenHenning committed:
Remove KitKat support.

This commit does a bunch of CI and build clean-up, but its primary set
of changes is to remove references and code corresponding to KitKat.

BenHenning reviewed (commented):
Self-reviewed.

BenHenning committed:
Merge branch 'develop' into remove-kitkat-support

BenHenning committed:
Merge branch 'develop' into remove-kitkat-support

BenHenning committed:
Fix broken test post KitKat removal.

BenHenning reviewed (commented):
Self-reviewed latest change.

BenHenning commented:
PTAL @adhiamboperes. This should be ready for review now--there was only 1 failing test and it's been fixed. I'll double check the CI run when it completes, but I expect everything to pass now.

adhiamboperes reviewed (approved):
Thanks @BenHenning! This LGTM, although more coverage is required for NetworkModule.

BenHenning committed:
Merge branch 'develop' into remove-kitkat-support

BenHenning committed:
Add better coverage in NetworkModuleTest.

Also fix NetworkLoggingInterceptor.

BenHenning reviewed (commented):
Self-reviewed latest changes.

BenHenning commented:
> Thanks @BenHenning! This LGTM, although more coverage is required for NetworkModule.

Thanks @adhiamboperes! I've addressed this--PTAL since there are quite a few new tests being added (and some code generally being changed). I've also added 3 new bullet items to the PR description (which are copied below) if that context is helpful.

> * `NetworkModuleTest` has had its tests redesigned and better filled in in order to pass the code coverage minimum requirement (since old tests were removed). Note that it's not quite perfect in what it's testing, but it is at least verifying the complete configuration of the `Retrofit` instance, and the singleton properties of the two services currently supported (verifying that the services are set up would require a lot more testing that seems outside the direct scope of this test, so it seems okay to ignore here).
> * As part of the previous item's work, `NetworkLoggingInterceptor` had an issue fixed where it could cause an OkHttp exception to be thrown when trying to process logs (due to reading the response body twice). Reverting this change will now cause breakage failures in `NetworkModuleTest`.
> * `NetworkModule`'s initialization order was changed for slightly better network request logging (see the new comment in that file for more context).

@kkmurerwa I'm curious whether the last 2 bullet items above may be relevant to Retrofit calls not being logged in #5243? If there's an exception occurring during the actual call pipeline, it will fail before we even have an opportunity to log the issue (which is what I was observing in tests). Can you confirm whether the changes in this PR actually fix that issue?

@seanlip PTAL for codeowners change.

seanlip reviewed (approved):
Thanks @BenHenning, LGTM for codeowners.

kkmurerwa commented:
@BenHenning I will run the app on this branch and see if the retrofit issue is fixed.

BenHenning committed:
Merge branch 'develop' into remove-kitkat-support

adhiamboperes reviewed (approved):
Thanks @BenHenning, this LGTM.

kkmurerwa commented:
@BenHenning The issue is not fixed yet.

![image](https://github.com/user-attachments/assets/bdf0ead8-e9e4-4be9-b18e-f8943419eb01)

BenHenning commented:
> @BenHenning The issue is not fixed yet.
> 
> ![image](https://private-user-images.githubusercontent.com/18438114/413621755-bdf0ead8-e9e4-4be9-b18e-f8943419eb01.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk5MjE1NTgsIm5iZiI6MTczOTkyMTI1OCwicGF0aCI6Ii8xODQzODExNC80MTM2MjE3NTUtYmRmMGVhZDgtZTllNC00YmU5LWIxOGUtZjg5NDM0MTllYjAxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE4VDIzMjczOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhN2U0NjZjNmZiZTBiMDIxZWY1NzI5MDdmY2FmY2UzYjAwMjYxNmIyMTk4ZGMyMmU4Njk3ODlmNmFkN2MzOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wfA5edL0eGAwVRQDwJpMjUQKYgqqkmqr2eXFAcPq0Pk)

@kkmurerwa I don't think that stack trace actually indicates anything. See https://stackoverflow.com/a/59636900. What needs to be tested here is whether the job workers are actually correctly trying to connect to Oppia web via Retrofit after this PR's changes (since they would silently fail before due to the incorrect Retrofit configuration).

BenHenning commented:
Thanks @adhiamboperes and @seanlip for the reviews. Enabling auto-merge on this and removing the KitKat required checks.

Edit: Actually, there may not need to be any workflow renames. Waiting to see.

BenHenning committed:
Merge branch 'develop' into remove-kitkat-support

adhiamboperes commented:
@BenHenning, there seems to be a merge confict in the PR, PTAL.

BenHenning commented:
Ah. Will resolve that momentarily, thanks @adhiamboperes.

I also now see what needs to be done for CI. We removed the workflow for verifying the APK builds since those targets were removed. Yay for slightly more efficient CI!

BenHenning committed:
Merge branch 'develop' into remove-kitkat-support

BenHenning reviewed (commented):
Double checked the changes in `build_flavors.bzl` since there were conflicts to resolve from #5664. The resolution seemed to be clean and was simple, so I don't think another review check is needed here.
